Unbounded GC Thread Pool Expansion for CRIU Restore

RSalman opened the pull request:
Initial set of changes to expand the GC thread pool beyond the dispatcher initialization done at VM startup. This involves reallocating memory for the dispatcher tables, these tables store thread pointers/meta data and are sized based on the initial thread count at startup. Given the restore machine configuration, the thread pool may need to be expanded beyond the initial capacity, hence these tables must be reinited to accommodate a larger thread count.

Some things to note for unbounded expansion considering this capability is still under development:
- By default, the restore thread count is still capped based on startup conditions (i.e this capability is disabled). To enable this, a restore thread count must be explicitly set.
- Only expected to work with the standard collectors.
- Performance regressions are expected, no other known issues/crashes with standard collectors (GC components function as expected).
- Disabled for non-standard collectors. Balanced GC requires further work, Balanced GC components require reinit (ones tightly coupled to the # of GC threads). 

Signed-off-by: Salman Rana <salman.rana@ibm.com>

RSalman commented:
@amicic @dmitripivkine please review

cc @@tajila

RSalman commented:
Simplified computation of GC thread count, `_extensions->gcThreadCount` is now inited using exisiting `initializeGCThreadCount` call prior to dispatcher pool expand call. This is can be done using  `_delegate.setMaxGCThreadCount(env, dispatcher->getPoolMaxCapacity());` after checkpoint contract,  `initializeGCThreadCount` will account for pool size.

amicic commented on gc/base/Configuration.cpp:
```diff
@@ -511,16 +511,22 @@ MM_Configuration::createParallelDispatcher(MM_EnvironmentBase *env, omrsig_handl
 
-	extensions->dispatcher->contractThreadPool(env, extensions->checkpointGCthreadCount);
+	dispatcher->contractThreadPool(env, extensions->checkpointGCthreadCount);
+	_delegate.setMaxGCThreadCount(env, dispatcher->getPoolMaxCapacity());
```
add a comment, that this line is temporary, to prevent growth thread pool beyond the one at snapshot side till more dependent functional components (beyond just Dispatcher) are ensured to be ok with the growth

amicic commented on gc/base/ParallelDispatcher.cpp:
```diff
@@ -757,4 +754,61 @@ MM_ParallelDispatcher::expandThreadPool(MM_EnvironmentBase *env)
 }
+
+bool
+MM_ParallelDispatcher::reinitialize(MM_EnvironmentBase *env, uintptr_t newPoolSize)
```
append the name so it's more descriptive (that it's about the thread table/pool)

amicic commented on gc/base/ParallelDispatcher.hpp:
```diff
@@ -84,7 +84,7 @@ class MM_ParallelDispatcher : public MM_BaseVirtual
 
 #if defined(J9VM_OPT_CRIU_SUPPORT)
-	uintptr_t _threadCountMaximumAtCheckpointStartup;
+	uintptr_t _poolMaxCapacity;
```
add a short comment (refer to the tables)

amicic commented on gc/base/ParallelDispatcher.cpp:
```diff
@@ -723,20 +723,17 @@ MM_ParallelDispatcher::expandThreadPool(MM_EnvironmentBase *env)
+			newThreadCount, _poolMaxCapacity,
+			_extensions->configuration->supportedGCThreadCount(env), preExpandThreadCount);
 
+	if ((newThreadCount > preExpandThreadCount) && (result = reinitialize(env, newThreadCount))) {
```
do the table reinitialize as an explicit step ahead of this thread count check

amicic replied:
consider moving the thread count check inside internalStartupThreads - it would be consistent with how/when reinitialize for tables does the check

RSalman replied:
We'll need this check in this method,`internalStartupThreads` does the check already `	while (workerThreadCount < maxWorkerThreadIndex) ` but its return value is based on wether it failed or succeeded in starting the threads independent if check fails or passes. We need the check to set the threadCount vars aside from calling `internalStartupThreads`

amicic commented on gc/base/ParallelDispatcher.cpp:
```diff
@@ -757,4 +754,61 @@ MM_ParallelDispatcher::expandThreadPool(MM_EnvironmentBase *env)
+		/* Re-size/allocate the dispatcher tables: _threadTable, _statusTable & _taskTable. */
+		 omrthread_t *threadTableTemp;
+		 uintptr_t *statusTableTemp;
+		 MM_Task **taskTableTemp;
```
to comply with coding standards, into these with null, or move the declaration later at the time they are first initialized/used

amicic commented on gc/base/ParallelDispatcher.cpp:
```diff
@@ -757,4 +754,61 @@ MM_ParallelDispatcher::expandThreadPool(MM_EnvironmentBase *env)
+		 OMR::GC::Forge *forge = env->getForge();
+
+		 threadTableTemp = (omrthread_t *)forge->allocate(newPoolSize * sizeof(omrthread_t), OMR::GC::AllocationCategory::FIXED, OMR_GET_CALLSITE());
+		 if(!threadTableTemp) {
```
put space between 'if' and '('

RSalman commented:
In addition to addressing the review comments, I added an adjustment to ensure that the restore TC does not fall below the checkpoint TC. This was the behaviour when TC with inited in `expandTheradPool`

``` C++ 
	uintptr_t checkpointThreadCount = extensions->gcThreadCount;

	initializeGCThreadCount(env);

	/* Re-inited thread count should not be lower than the thread count at checkpoint. */
	extensions->gcThreadCount = OMR_MAX(checkpointThreadCount, extensions->gcThreadCount);

	return extensions->dispatcher->expandThreadPool(env);
```

amicic commented on gc/base/ParallelDispatcher.cpp:
```diff
@@ -723,20 +723,21 @@ MM_ParallelDispatcher::expandThreadPool(MM_EnvironmentBase *env)
-	if (newThreadCount > preExpandThreadCount) {
+	result = reinitializeThreadPool(env, newThreadCount);
 
+	if ((newThreadCount > preExpandThreadCount) && result) {
```
minor (what ever you like) but consider doing as a pair of nested ifs:
if (result = reinitialize())
    if (newThreadCount > preExpanded)
    
the most important for me was that Thread pool  was a explicit step addressed before considering that threads themselves should be addressed

amicic commented on gc/base/ParallelDispatcher.cpp:
```diff
@@ -757,4 +758,57 @@ MM_ParallelDispatcher::expandThreadPool(MM_EnvironmentBase *env)
+		 memset(threadTableTemp, 0, newPoolSize * sizeof(omrthread_t));
+
+		 uintptr_t *statusTableTemp = (uintptr_t *)forge->allocate(newPoolSize * sizeof(uintptr_t *), OMR::GC::AllocationCategory::FIXED, OMR_GET_CALLSITE());
+		 if(!statusTableTemp) {
```
put space between 'if' and '('

amicic commented:
while I approved this, I realized this should still hold a bit, till a downstream change is merged

babsingh reviewed (changes requested):
The commit message and PR description do not match?

RSalman commented:
> The commit message and PR description do not match?

Fixed. 

I'll be opening an issue to track the work around CRIU GC, it will cover the pending work mentioned in the comments.

babsingh commented on gc/base/Configuration.cpp:
```diff
@@ -511,16 +511,33 @@ MM_Configuration::createParallelDispatcher(MM_EnvironmentBase *env, omrsig_handl
-	extensions->dispatcher->contractThreadPool(env, extensions->checkpointGCthreadCount);
+	dispatcher->contractThreadPool(env, extensions->checkpointGCthreadCount);
+
+	/* This line is temporary, it's required to prevent thread pool growth beyond the initial pool sizing done at startup.
+	 * It's required until dependent functional components (beyond just Dispatcher) are ensured to be ok with the growth.
+	 */
```
- exceeds more than 100 chars, too long.
- `ok` should be replaced with a better (technical) phrase.

RSalman replied:
👍

babsingh commented on gc/base/ParallelDispatcher.cpp:
```diff
@@ -757,4 +760,66 @@ MM_ParallelDispatcher::expandThreadPool(MM_EnvironmentBase *env)
+		_threadTable = threadTableTemp;
+
+		_poolMaxCapacity = newPoolSize;
+	}
```
is it correct to return true for the `else` case (`newPoolSize` <= `_poolMaxCapacity`)?

RSalman replied:
Yup, I've updated the documentation for this function to explain:

```bool indicating if the thread pool tables can accommodate the newPoolSize (i.e., successfully reallocated or already have capacity). * ```

babsingh commented:
Previously unaddressed: 

> The commit message and PR description should match.

RSalman commented:
> Previously unaddressed:
> The commit message and PR description should match.

I've taken a look and not sure what the diff between the two is, could you please specify.

RSalman commented on gc/base/Configuration.cpp:
```diff
@@ -511,16 +511,33 @@ MM_Configuration::createParallelDispatcher(MM_EnvironmentBase *env, omrsig_handl
+
+	/* Currently, Threads don't shutdown during restore, so ensure thread count doesn't fall below the checkpoint thread count.
+	 * This adjustment can be removed in the future when dispatcher thread shutdown is sufficiently tested at restore.
+	 */
```
Fixed, it shouldn't be capital

babsingh commented:
> I've taken a look and not sure what the diff between the two is, could you please specify.

My bad, they are the same. I expected the PR description to change. But the commit message was updated. I didn't check the commit message when posting the above comment.

babsingh commented:
jenkins build all
